lib/Federation/Notifications.php

```diff
@@ -0,0 +1,254 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Federation;
+
+use OCA\FederatedFileSharing\AddressHandler;
+use OCA\Talk\AppInfo\Application;
+use OCA\Talk\BackgroundJob\RetryJob;
+use OCA\Talk\MatterbridgeManager;
+use OCA\Talk\Model\Attendee;
+use OCA\Talk\Model\AttendeeMapper;
+use OCA\Talk\Participant;
+use OCA\Talk\Room;
+use OCP\BackgroundJob\IJobList;
+use OCP\DB\Exception;
+use OCP\Federation\ICloudFederationFactory;
+use OCP\Federation\ICloudFederationNotification;
+use OCP\Federation\ICloudFederationProviderManager;
+use OCP\IUser;
+use OCP\IUserManager;
+use Psr\Log\LoggerInterface;
+
+class Notifications {
+	/** @var ICloudFederationFactory */
+	private $cloudFederationFactory;
+
+	/** @var AddressHandler */
+	private $addressHandler;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var ICloudFederationProviderManager */
+	private $federationProviderManager;
+
+	/** @var IJobList */
+	private $jobList;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var AttendeeMapper */
+	private $attendeeMapper;
+
+	/** @var MatterbridgeManager */
+	private $matterbridgeManager;
+
+	public function __construct(
+		ICloudFederationFactory $cloudFederationFactory,
+		AddressHandler $addressHandler,
+		LoggerInterface $logger,
+		ICloudFederationProviderManager $federationProviderManager,
+		IJobList $jobList,
+		IUserManager $userManager,
+		AttendeeMapper $attendeeMapper,
+		MatterbridgeManager $matterbridgeManager
+	) {
+		$this->cloudFederationFactory = $cloudFederationFactory;
+		$this->addressHandler = $addressHandler;
+		$this->logger = $logger;
+		$this->federationProviderManager = $federationProviderManager;
+		$this->jobList = $jobList;
+		$this->userManager = $userManager;
+		$this->attendeeMapper = $attendeeMapper;
+		$this->matterbridgeManager = $matterbridgeManager;
+	}
+
+	public function sendRemoteShare(string $providerId, string $token, string $shareWith, string $sharedBy,
+									string $sharedByFederatedId, string $shareType, Room $room): bool {
+		[$user, $remote] = $this->addressHandler->splitUserRemote($shareWith);
+
+		$roomName = $room->getName();
+		$roomType = $room->getType();
+		$roomToken = $room->getToken();
+
+		if (!($user && $remote)) {
+			$this->logger->info(
+				"could not share $roomToken, invalid contact $shareWith",
+				['app' => Application::APP_ID]
+			);
+			return false;
+		}
+
+		/** @var IUser|null $roomOwner */
+		$roomOwner = null;
+		try {
+			$roomOwners = $this->attendeeMapper->getActorsByParticipantTypes($room->getId(), [Participant::OWNER]);
+			if (!empty($roomOwners) && $roomOwners[0]->getActorType() === Attendee::ACTOR_USERS) {
+				$roomOwner = $this->userManager->get($roomOwners[0]->getActorId());
+			}
+		} catch (Exception $e) {
+			$roomOwner = $this->matterbridgeManager->getBotUser($room);
```

Comment:
I would just pick the first moderator if there is no owner. A local moderator is always needed, the last one leaving the chat is blocked and needs to delete the conversation instead of leaving it.